File: 0090-Subsets-WithDup/0090.js
// 90. 子集 II
// 给你一个整数数组 nums ，其中可能包含重复元素，请你返回该数组所有可能的子集（幂集）。

// 解集 不能 包含重复的子集。返回的解集中，子集可以按 任意顺序 排列。



// 示例 1：

// 输入：nums = [1, 2, 2]
// 输出：[[], [1], [1, 2], [1, 2, 2], [2], [2, 2]]
// 示例 2：

// 输入：nums = [0]
// 输出：[[], [0]]


// 提示：

// 1 <= nums.length <= 10
//   - 10 <= nums[i] <= 10

var subsetsWithDup = function (nums) {
  nums.sort((a, b) => a - b);
  let t = [], ans = [];
  const n = nums.length;
  for (let mask = 0; mask < (1 << n); ++mask) {
    t = [];
    let flag = true;
    for (let i = 0; i < n; ++i) {
      if ((mask & (1 << i)) != 0) {
        if (i > 0 && (mask >> (i - 1) & 1) == 0 && nums[i] == nums[i - 1]) {
          flag = false;
          break;
        }
        t.push(nums[i]);
      }
    }
    if (flag) {
      ans.push(t.slice());
    }
  }
  return ans;
}

File: 0-100/0090-Subsets-WithDup/0090.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        t = []

        def dfs(choosePre: bool, cur: int):
            if cur == len(nums):
                ans.append(t[:])
                return
            dfs(False, cur + 1)
            if not choosePre and cur > 0 and nums[cur] == nums[cur - 1]:
                return
            t.append(nums[cur])
            dfs(True, cur + 1)
            t.pop()
        dfs(False, 0)
        return ans